src/models/history/tree-manager.ts

```diff
@@ -57,12 +77,137 @@ export const TreeManager = types
     return self.activeHistoryEntries.find(entry => entry.id === historyEntryId);
   }
 }))
+.actions(self => ({
+  completeHistoryEntry(entry: Instance<typeof HistoryEntry>) {
+    entry.state = "complete";
+
+    // Remove the entry from the activeHistoryEntries
+    detach(entry);
+
+    // If the history entry resulted in no changes don't add it to the real
+    // history.
+    //
+    // TODO: we might not want to do this, it could be useful for researchers
+    // and teachers to see actions the student took even if they didn't change
+    // state. For example if a button was clicked even if it didn't change the
+    // state we might want to show that somehow. There are lots of entries
+    // that are empty though, so they are removed for the time being.
+    if (entry.records.length === 0) {
+      destroy(entry);
+      return;
+    }
+
+    // Re-attach the entry to the history 
+    self.document.history.push(entry);
+
+    // Add the entry to the undo stack if it is undoable.
+    //
+    // TODO: Is it best to wait until the entry is complete like this? It
+    // might be better to add it earlier so it has the right position in the
+    // undo stack. For example if a user action caused some async behavior
+    // that takes a while, should its place in the stack be at the beginning
+    // or end of these changes? As a downside, if we add it earlier the undo
+    // stack will have incomplete entries in sometimes.
+    if (entry.undoable) {
+      self.undoStore.addHistoryEntry(entry);
+    }
+
+    // TODO: send this history entry to firestore
+
+    // Create the document in firestore if necessary
+    const validateCommentableDocument = 
+      getFirebaseFunction<ICommentableDocumentParams>("validateCommentableDocument_v1");
```

Comment:
Not for this PR, but we should rename this function to something like `validateFirestoreDocumentMetadata_v1`.

Reply:
I added a FIXME for this. I'm still leaning towards removing the use of the cloud function.